File: packages/py/src/qorrdk/presets.py
"""The five preset profiles and their documented default fields.

These mirror the network's published profile table. The proof system for each
profile is the one its settlement paradigm requires (optimistic -> fraud,
zk -> snark, based -> none).
"""

from __future__ import annotations

from typing import Callable

from .config.builder import RollupConfigBuilder
from .config.types import RollupConfig, SequencerParams
from .constants import DEFAULT_CHALLENGE_BOND_UQOR, DEFAULT_RDK_PARAMS
from .enums import (
    DABackend,
    GasModel,
    ProfileName,
    ProofSystem,
    SequencerMode,
    SettlementParadigm,
    VmType,
)

_DEFAULT_CHALLENGE_WINDOW = int(DEFAULT_RDK_PARAMS["default_challenge_window"])

#: Resolved preset fields, keyed by profile. Each maps to the keyword arguments
#: for a :class:`RollupConfig` (minus ``rollup_id`` and ``stake_amount_uqor``).
PRESET_DEFAULTS: dict[ProfileName, dict] = {
    ProfileName.DEFI: {
        "profile": ProfileName.DEFI,
        "settlement": SettlementParadigm.ZK,
        "sequencer": SequencerMode.DEDICATED,
        "da": DABackend.NATIVE,
        "proof_system": ProofSystem.SNARK,
        "gas_model": GasModel.EIP1559,
        "vm_type": VmType.EVM,
        "block_time_ms": 500,
        "max_tx_per_block": 10000,
    },
    ProfileName.GAMING: {
        "profile": ProfileName.GAMING,
        "settlement": SettlementParadigm.BASED,
        "sequencer": SequencerMode.BASED,
        "da": DABackend.NATIVE,
        "proof_system": ProofSystem.NONE,
        "gas_model": GasModel.FLAT,
        "vm_type": VmType.CUSTOM,
        "block_time_ms": 200,
        "max_tx_per_block": 50000,
    },
    ProfileName.NFT: {
        "profile": ProfileName.NFT,
        "settlement": SettlementParadigm.OPTIMISTIC,
        "sequencer": SequencerMode.DEDICATED,
        "da": DABackend.CELESTIA,
        "proof_system": ProofSystem.FRAUD,
        "gas_model": GasModel.STANDARD,
        "vm_type": VmType.COSMWASM,
        "block_time_ms": 2000,
        "max_tx_per_block": 5000,
        "challenge_window_secs": _DEFAULT_CHALLENGE_WINDOW,
        "challenge_bond_uqor": DEFAULT_CHALLENGE_BOND_UQOR,
    },
    ProfileName.ENTERPRISE: {
        "profile": ProfileName.ENTERPRISE,
        "settlement": SettlementParadigm.BASED,
        "sequencer": SequencerMode.BASED,
        "da": DABackend.NATIVE,
        "proof_system": ProofSystem.NONE,
        "gas_model": GasModel.SUBSIDIZED,
        "vm_type": VmType.EVM,
        "block_time_ms": 1000,
        "max_tx_per_block": 20000,
    },
    ProfileName.CUSTOM: {
        "profile": ProfileName.CUSTOM,
        "settlement": SettlementParadigm.OPTIMISTIC,
        "sequencer": SequencerMode.DEDICATED,
        "da": DABackend.NATIVE,
        "proof_system": ProofSystem.FRAUD,
        "gas_model": GasModel.STANDARD,
        "vm_type": VmType.EVM,
        "block_time_ms": 1000,
        "max_tx_per_block": 10000,
        "challenge_window_secs": _DEFAULT_CHALLENGE_WINDOW,
        "challenge_bond_uqor": DEFAULT_CHALLENGE_BOND_UQOR,
    },
}


def _make_preset(name: ProfileName) -> Callable[..., RollupConfigBuilder]:
    def preset(rollup_id: str = "", **overrides: object) -> RollupConfigBuilder:
        base = dict(PRESET_DEFAULTS[name])
        sequencer_params = overrides.pop("sequencer_params", None)
        if isinstance(sequencer_params, dict):
            sequencer_params = SequencerParams(**sequencer_params)
        base.update(overrides)
        base["rollup_id"] = rollup_id
        base["profile"] = name
        config = RollupConfig(sequencer_params=sequencer_params, **base)
        return RollupConfigBuilder(config)

    preset.__name__ = f"preset_{name.value}"
    preset.__doc__ = (
        f"Build a RollupConfigBuilder for the '{name.value}' profile, applying any "
        "keyword overrides on top of the profile's documented defaults."
    )
    return preset


class _Presets:
    """First-class preset profiles. Each returns a :class:`RollupConfigBuilder`."""

    defi = staticmethod(_make_preset(ProfileName.DEFI))
    gaming = staticmethod(_make_preset(ProfileName.GAMING))
    nft = staticmethod(_make_preset(ProfileName.NFT))
    enterprise = staticmethod(_make_preset(ProfileName.ENTERPRISE))
    custom = staticmethod(_make_preset(ProfileName.CUSTOM))


#: Singleton holding the five preset builders.
presets = _Presets()


__all__ = ["PRESET_DEFAULTS", "presets"]

File: packages/jvm/src/main/java/io/github/qorechain/rdk/presets/Presets.java
package io.github.qorechain.rdk.presets;

import io.github.qorechain.rdk.config.Constants;
import io.github.qorechain.rdk.config.Enums.DABackend;
import io.github.qorechain.rdk.config.Enums.GasModel;
import io.github.qorechain.rdk.config.Enums.ProfileName;
import io.github.qorechain.rdk.config.Enums.ProofSystem;
import io.github.qorechain.rdk.config.Enums.SequencerMode;
import io.github.qorechain.rdk.config.Enums.SettlementParadigm;
import io.github.qorechain.rdk.config.Enums.VmType;
import io.github.qorechain.rdk.config.RollupConfig;
import io.github.qorechain.rdk.config.RollupConfigBuilder;
import java.util.EnumMap;
import java.util.Map;
import java.util.function.Consumer;

/**
 * The five preset profiles and their documented default fields.
 *
 * <p>These mirror the network's published profile table. The proof system for each profile is the
 * one its settlement paradigm requires (optimistic → fraud, zk → snark, based → none).
 */
public final class Presets {
    private Presets() {}

    /** The documented default fields for each profile (excluding rollupId and stake). */
    public static final Map<ProfileName, RollupConfig> PRESET_DEFAULTS;

    static {
        Map<ProfileName, RollupConfig> m = new EnumMap<>(ProfileName.class);

        RollupConfig defi = base(ProfileName.DEFI);
        defi.settlement = SettlementParadigm.ZK;
        defi.sequencer = SequencerMode.DEDICATED;
        defi.da = DABackend.NATIVE;
        defi.proofSystem = ProofSystem.SNARK;
        defi.gasModel = GasModel.EIP1559;
        defi.vmType = VmType.EVM;
        defi.blockTimeMs = 500;
        defi.maxTxPerBlock = 10000;
        m.put(ProfileName.DEFI, defi);

        RollupConfig gaming = base(ProfileName.GAMING);
        gaming.settlement = SettlementParadigm.BASED;
        gaming.sequencer = SequencerMode.BASED;
        gaming.da = DABackend.NATIVE;
        gaming.proofSystem = ProofSystem.NONE;
        gaming.gasModel = GasModel.FLAT;
        gaming.vmType = VmType.CUSTOM;
        gaming.blockTimeMs = 200;
        gaming.maxTxPerBlock = 50000;
        m.put(ProfileName.GAMING, gaming);

        RollupConfig nft = base(ProfileName.NFT);
        nft.settlement = SettlementParadigm.OPTIMISTIC;
        nft.sequencer = SequencerMode.DEDICATED;
        nft.da = DABackend.CELESTIA;
        nft.proofSystem = ProofSystem.FRAUD;
        nft.gasModel = GasModel.STANDARD;
        nft.vmType = VmType.COSMWASM;
        nft.blockTimeMs = 2000;
        nft.maxTxPerBlock = 5000;
        nft.challengeWindowSecs = Constants.DEFAULT_CHALLENGE_WINDOW_SECS;
        nft.challengeBondUqor = Constants.DEFAULT_CHALLENGE_BOND_UQOR;
        m.put(ProfileName.NFT, nft);

        RollupConfig enterprise = base(ProfileName.ENTERPRISE);
        enterprise.settlement = SettlementParadigm.BASED;
        enterprise.sequencer = SequencerMode.BASED;
        enterprise.da = DABackend.NATIVE;
        enterprise.proofSystem = ProofSystem.NONE;
        enterprise.gasModel = GasModel.SUBSIDIZED;
        enterprise.vmType = VmType.EVM;
        enterprise.blockTimeMs = 1000;
        enterprise.maxTxPerBlock = 20000;
        m.put(ProfileName.ENTERPRISE, enterprise);

        RollupConfig custom = base(ProfileName.CUSTOM);
        custom.settlement = SettlementParadigm.OPTIMISTIC;
        custom.sequencer = SequencerMode.DEDICATED;
        custom.da = DABackend.NATIVE;
        custom.proofSystem = ProofSystem.FRAUD;
        custom.gasModel = GasModel.STANDARD;
        custom.vmType = VmType.EVM;
        custom.blockTimeMs = 1000;
        custom.maxTxPerBlock = 10000;
        custom.challengeWindowSecs = Constants.DEFAULT_CHALLENGE_WINDOW_SECS;
        custom.challengeBondUqor = Constants.DEFAULT_CHALLENGE_BOND_UQOR;
        m.put(ProfileName.CUSTOM, custom);

        PRESET_DEFAULTS = m;
    }

    private static RollupConfig base(ProfileName profile) {
        RollupConfig c = new RollupConfig();
        c.profile = profile;
        return c;
    }

    /**
     * Build a {@link RollupConfigBuilder} for a profile, applying any overrides on top of the
     * profile's documented defaults. The {@code profile} field always reflects the chosen preset.
     */
    public static RollupConfigBuilder preset(ProfileName name, Consumer<RollupConfig> overrides) {
        RollupConfig config = PRESET_DEFAULTS.get(name).copy();
        config.rollupId = "";
        if (overrides != null) {
            overrides.accept(config);
        }
        config.profile = name;
        return new RollupConfigBuilder(config);
    }

    public static RollupConfigBuilder defi(Consumer<RollupConfig> overrides) {
        return preset(ProfileName.DEFI, overrides);
    }

    public static RollupConfigBuilder gaming(Consumer<RollupConfig> overrides) {
        return preset(ProfileName.GAMING, overrides);
    }

    public static RollupConfigBuilder nft(Consumer<RollupConfig> overrides) {
        return preset(ProfileName.NFT, overrides);
    }

    public static RollupConfigBuilder enterprise(Consumer<RollupConfig> overrides) {
        return preset(ProfileName.ENTERPRISE, overrides);
    }

    public static RollupConfigBuilder custom(Consumer<RollupConfig> overrides) {
        return preset(ProfileName.CUSTOM, overrides);
    }
}
